// puregram/lib/common/media-group.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaGroup = void 0;
const inspectable_1 = require("inspectable");
/** This object represent a media group: a group of contexts with some attachments in it */
class MediaGroup {
    constructor(options) {
        this.options = options;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    /** Returns media group's ID */
    get id() {
        return this.options.id;
    }
    /** Returns a list of contexts (constructed earlier) every single of which contains an attachment */
    get contexts() {
        return this.options.contexts;
    }
    /** Returns a list of attachments (mapped through `contexts`) */
    get attachments() {
        const attachments = this.contexts.map(context => context.attachments).flat();
        return attachments;
    }
}
exports.MediaGroup = MediaGroup;
(0, inspectable_1.inspectable)(MediaGroup, {
    serialize(mg) {
        return {
            id: mg.id,
            contexts: mg.contexts,
            attachments: mg.attachments
        };
    }
});
