// puregram/lib/common/attachments/document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocumentAttachment = void 0;
const inspectable_1 = require("inspectable");
const structures_1 = require("../structures");
const file_attachment_1 = require("./file-attachment");
/**
 * This object represents a general file (as opposed to photos, voice messages
 * and audio files).
 */
class DocumentAttachment extends file_attachment_1.FileAttachment {
    constructor() {
        super(...arguments);
        this.attachmentType = 'document';
    }
    /** Document thumbnail as defined by sender */
    get thumbnail() {
        const { thumbnail } = this.payload;
        if (!thumbnail) {
            return;
        }
        return new structures_1.PhotoSize(thumbnail);
    }
    /** Original filename as defined by sender */
    get fileName() {
        return this.payload.file_name;
    }
    /** MIME type of the file as defined by sender */
    get mimeType() {
        return this.payload.mime_type;
    }
    /** File size */
    get fileSize() {
        return this.payload.file_size;
    }
    toJSON() {
        var _a;
        return {
            file_id: this.fileId,
            file_unique_id: this.fileUniqueId,
            thumbnail: (_a = this.thumbnail) === null || _a === void 0 ? void 0 : _a.toJSON(),
            file_name: this.fileName,
            mime_type: this.mimeType,
            file_size: this.fileSize
        };
    }
}
exports.DocumentAttachment = DocumentAttachment;
(0, inspectable_1.inspectable)(DocumentAttachment, {
    serialize(attachment) {
        return {
            fileId: attachment.fileId,
            fileUniqueId: attachment.fileUniqueId,
            thumbnail: attachment.thumbnail,
            fileName: attachment.fileName,
            mimeType: attachment.mimeType,
            fileSize: attachment.fileSize
        };
    }
});
