// puregram/lib/common/attachments/photo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhotoAttachment = void 0;
const inspectable_1 = require("inspectable");
const attachment_1 = require("./attachment");
/** This object represents a photo file with it's sizes */
class PhotoAttachment extends attachment_1.Attachment {
    constructor(payload) {
        super();
        this.attachmentType = 'photo';
        this.payload = payload;
        this.sorted = [...this.payload].sort((first, second) => (second.width * second.height) - (first.width * first.height));
    }
    /** Photo sizes */
    get sizes() {
        return this.payload;
    }
    /** Biggest size of the photo */
    get bigSize() {
        return this.sorted[0];
    }
    /** Medium size of the photo */
    get mediumSize() {
        return this.sorted[Math.floor(this.sorted.length / 2)];
    }
    /** Smallest size of the photo */
    get smallSize() {
        return this.sorted[this.sorted.length - 1];
    }
    toJSON() {
        return [
            this.smallSize.toJSON(),
            this.mediumSize.toJSON(),
            this.bigSize.toJSON()
        ];
    }
}
exports.PhotoAttachment = PhotoAttachment;
(0, inspectable_1.inspectable)(PhotoAttachment, {
    serialize(attachment) {
        return {
            smallSize: attachment.smallSize,
            mediumSize: attachment.mediumSize,
            bigSize: attachment.bigSize
        };
    }
});
