// puregram/lib/common/attachments/poll.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PollAttachment = void 0;
const helpers_1 = require("../../utils/helpers");
const poll_1 = require("../structures/poll");
const attachment_1 = require("./attachment");
class PollAttachment extends poll_1.Poll {
    constructor() {
        super(...arguments);
        this.attachmentType = 'poll';
    }
    toJSON() {
        var _a;
        return {
            id: this.id,
            question: this.question,
            options: this.options.map(option => option.toJSON()),
            total_voter_count: this.totalVoterCount,
            is_closed: this.isClosed(),
            is_anonymous: this.isAnonymous(),
            type: this.type,
            allows_multiple_answers: this.allowsMultipleAnswers,
            correct_option_id: this.correctOptionId,
            explanation: this.explanation,
            explanation_entities: (_a = this.explanationEntities) === null || _a === void 0 ? void 0 : _a.map(entity => entity.toJSON()),
            open_period: this.openPeriod,
            close_date: this.closeDate
        };
    }
}
exports.PollAttachment = PollAttachment;
(0, helpers_1.applyMixins)(PollAttachment, [attachment_1.Attachment]);
