// puregram/lib/common/attachments/video.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VideoAttachment = void 0;
const inspectable_1 = require("inspectable");
const structures_1 = require("../structures");
const file_attachment_1 = require("./file-attachment");
/** This object represents a video file. */
class VideoAttachment extends file_attachment_1.FileAttachment {
    constructor() {
        super(...arguments);
        this.attachmentType = 'video';
    }
    /** Video width as defined by sender */
    get width() {
        return this.payload.width;
    }
    /** Video height as defined by sender */
    get height() {
        return this.payload.height;
    }
    /** Duration of the video in seconds as defined by sender */
    get duration() {
        return this.payload.duration;
    }
    /** Video thumbnail */
    get thumbnail() {
        const { thumbnail } = this.payload;
        if (!thumbnail) {
            return;
        }
        return new structures_1.PhotoSize(thumbnail);
    }
    /** Original filename as defined by sender */
    get fileName() {
        return this.payload.file_name;
    }
    /** Mime type of a file as defined by sender */
    get mimeType() {
        return this.payload.mime_type;
    }
    /** File size */
    get fileSize() {
        return this.payload.file_size;
    }
    toJSON() {
        var _a;
        return {
            file_id: this.fileId,
            file_unique_id: this.fileUniqueId,
            width: this.width,
            height: this.height,
            duration: this.duration,
            thumbnail: (_a = this.thumbnail) === null || _a === void 0 ? void 0 : _a.toJSON(),
            file_name: this.fileName,
            mime_type: this.mimeType,
            file_size: this.fileSize
        };
    }
}
exports.VideoAttachment = VideoAttachment;
(0, inspectable_1.inspectable)(VideoAttachment, {
    serialize(attachment) {
        return {
            fileId: attachment.fileId,
            fileUniqueId: attachment.fileUniqueId,
            width: attachment.width,
            height: attachment.height,
            duration: attachment.duration,
            thumbnail: attachment.thumbnail,
            fileName: attachment.fileName,
            mimeType: attachment.mimeType,
            fileSize: attachment.fileSize
        };
    }
});
