// puregram/lib/common/attachments/video-note.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VideoNoteAttachment = void 0;
const inspectable_1 = require("inspectable");
const structures_1 = require("../structures");
const file_attachment_1 = require("./file-attachment");
/** This object represents a video message. */
class VideoNoteAttachment extends file_attachment_1.FileAttachment {
    constructor() {
        super(...arguments);
        this.attachmentType = 'video_note';
    }
    /**
     * Video width and height (diameter of the video message) as defined by
     * sender
     */
    get length() {
        return this.payload.length;
    }
    /** Duration of the video in seconds as defined by sender */
    get duration() {
        return this.payload.duration;
    }
    /** Video thumbnail */
    get thumbnail() {
        const { thumbnail } = this.payload;
        if (!thumbnail) {
            return;
        }
        return new structures_1.PhotoSize(thumbnail);
    }
    /** File size */
    get fileSize() {
        return this.payload.file_size;
    }
    toJSON() {
        var _a;
        return {
            file_id: this.fileId,
            file_unique_id: this.fileUniqueId,
            length: this.length,
            duration: this.duration,
            thumbnail: (_a = this.thumbnail) === null || _a === void 0 ? void 0 : _a.toJSON(),
            file_size: this.fileSize
        };
    }
}
exports.VideoNoteAttachment = VideoNoteAttachment;
(0, inspectable_1.inspectable)(VideoNoteAttachment, {
    serialize(attachment) {
        return {
            fileId: attachment.fileId,
            fileUniqueId: attachment.fileUniqueId,
            length: attachment.length,
            duration: attachment.duration,
            thumbnail: attachment.thumbnail,
            fileSize: attachment.fileSize
        };
    }
});
