// puregram/lib/common/attachments/voice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VoiceAttachment = void 0;
const inspectable_1 = require("inspectable");
const file_attachment_1 = require("./file-attachment");
/** This object represents a voice note. */
class VoiceAttachment extends file_attachment_1.FileAttachment {
    constructor() {
        super(...arguments);
        this.attachmentType = 'voice';
    }
    /** Duration of the audio in seconds as defined by sender */
    get duration() {
        return this.payload.duration;
    }
    /** MIME type of the file as defined by sender */
    get mimeType() {
        return this.payload.mime_type;
    }
    /** File size */
    get fileSize() {
        return this.payload.file_size;
    }
    toJSON() {
        return {
            file_id: this.fileId,
            file_unique_id: this.fileUniqueId,
            duration: this.duration,
            mime_type: this.mimeType,
            file_size: this.fileSize
        };
    }
}
exports.VoiceAttachment = VoiceAttachment;
(0, inspectable_1.inspectable)(VoiceAttachment, {
    serialize(attachment) {
        return {
            fileId: attachment.fileId,
            fileUniqueId: attachment.fileUniqueId,
            duration: attachment.duration,
            mimeType: attachment.mimeType,
            fileSize: attachment.fileSize
        };
    }
});
