// puregram/lib/contexts/channel-chat-created.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelChatCreatedContext = void 0;
const inspectable_1 = require("inspectable");
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const mixins_1 = require("./mixins");
class ChannelChatCreatedContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'channel_chat_created',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
    /** Unique message identifier inside this chat */
    get id() {
        return this.payload.message_id;
    }
    /** Date the message was sent in Unix time */
    get createdAt() {
        return this.payload.date;
    }
}
exports.ChannelChatCreatedContext = ChannelChatCreatedContext;
(0, helpers_1.applyMixins)(ChannelChatCreatedContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(ChannelChatCreatedContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType
        };
    }
});
