// puregram/lib/contexts/chat-join-request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatJoinRequestContext = void 0;
const inspectable_1 = require("inspectable");
const helpers_1 = require("../utils/helpers");
const structures_1 = require("../common/structures");
const context_1 = require("./context");
const mixins_1 = require("./mixins");
class ChatJoinRequestContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'chat_join_request',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
}
exports.ChatJoinRequestContext = ChatJoinRequestContext;
(0, helpers_1.applyMixins)(ChatJoinRequestContext, [structures_1.ChatJoinRequest, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(ChatJoinRequestContext, {
    serialize(context) {
        return {
            chat: context.chat,
            from: context.from,
            date: context.date,
            bio: context.bio,
            inviteLink: context.inviteLink
        };
    }
});
