// puregram/lib/contexts/chat-shared.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatSharedContext = void 0;
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const inspectable_1 = require("inspectable");
const mixins_1 = require("./mixins");
/** This object contains information about the chat whose identifier was shared with the bot using a `KeyboardButtonRequestChat` button. */
class ChatSharedContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'chat_shared',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
        this.event = this.payload.chat_shared;
    }
    /** Identifier of the request */
    get requestId() {
        return this.event.request_id;
    }
    /** Identifier of the shared chat. This number may have more than 32 significant bits and some programming languages may have difficulty/silent defects in interpreting it. But it has at most 52 significant bits, so a 64-bit integer or double-precision float type are safe for storing this identifier. The bot may not have access to the chat and could be unable to use this identifier, unless the chat is already known to the bot by some other means. */
    get sharedChatId() {
        return this.event.chat_id;
    }
}
exports.ChatSharedContext = ChatSharedContext;
(0, helpers_1.applyMixins)(ChatSharedContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(ChatSharedContext, {
    serialize(context) {
        return {
            requestId: context.requestId,
            sharedChatId: context.sharedChatId
        };
    }
});
