// puregram/lib/contexts/chat-member.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatMemberContext = void 0;
const inspectable_1 = require("inspectable");
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const mixins_1 = require("./mixins");
class ChatMemberContext extends context_1.Context {
    constructor(options) {
        var _a;
        super({
            telegram: options.telegram,
            updateType: (_a = options.type) !== null && _a !== void 0 ? _a : 'chat_member',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
    /** Does this update have `invite_link` property? */
    hasInviteLink() {
        return this.inviteLink !== undefined;
    }
}
exports.ChatMemberContext = ChatMemberContext;
(0, helpers_1.applyMixins)(ChatMemberContext, [structures_1.ChatMemberUpdated, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(ChatMemberContext, {
    serialize(context) {
        return {
            senderId: context.senderId,
            chatId: context.chatId,
            chatType: context.chatType,
            oldChatMember: context.oldChatMember,
            newChatMember: context.newChatMember,
            date: context.date,
            inviteLink: context.inviteLink
        };
    }
});
