// puregram/lib/contexts/forum-topic-edited.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForumTopicEditedContext = void 0;
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const inspectable_1 = require("inspectable");
const mixins_1 = require("./mixins");
/** This object represents a service message about an edited forum topic. */
class ForumTopicEditedContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'forum_topic_edited',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
        this.event = this.payload.forum_topic_edited;
    }
    /** New name of the topic, if it was edited */
    get name() {
        return this.event.name;
    }
    /** Checks whether the `name` property has been edited */
    hasName() {
        return this.name !== undefined;
    }
    /** New identifier of the custom emoji shown as the topic icon, if it was edited; an empty string if the icon was removed */
    get iconCustomEmojiId() {
        return this.event.icon_custom_emoji_id;
    }
    /** Checks whether the `iconCustomEmojiId` property has been edited */
    hasIconCustomEmojiId() {
        return this.iconCustomEmojiId !== undefined;
    }
}
exports.ForumTopicEditedContext = ForumTopicEditedContext;
(0, helpers_1.applyMixins)(ForumTopicEditedContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(ForumTopicEditedContext, {
    serialize(context) {
        const payload = {
            name: context.name,
            iconCustomEmojiId: context.iconCustomEmojiId
        };
        return (0, helpers_1.filterPayload)(payload);
    }
});
