// puregram/lib/contexts/forum-topic-reopened.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForumTopicReopenedContext = void 0;
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const inspectable_1 = require("inspectable");
const mixins_1 = require("./mixins");
/** This object represents a service message about a forum topic reopened in the chat. Currently holds no information. */
class ForumTopicReopenedContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'forum_topic_reopened',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
}
exports.ForumTopicReopenedContext = ForumTopicReopenedContext;
(0, helpers_1.applyMixins)(ForumTopicReopenedContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(ForumTopicReopenedContext, {
    serialize(context) {
        return {};
    }
});
