// puregram/lib/contexts/general-forum-topic-hidden.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeneralForumTopicHiddenContext = void 0;
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const inspectable_1 = require("inspectable");
const mixins_1 = require("./mixins");
/** This object represents a service message about General forum topic hidden in the chat. Currently holds no information. */
class GeneralForumTopicHiddenContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'general_forum_topic_hidden',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
}
exports.GeneralForumTopicHiddenContext = GeneralForumTopicHiddenContext;
(0, helpers_1.applyMixins)(GeneralForumTopicHiddenContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(GeneralForumTopicHiddenContext, {
    serialize(context) {
        return {};
    }
});
