// puregram/lib/contexts/left-chat-member.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LeftChatMemberContext = void 0;
const inspectable_1 = require("inspectable");
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const mixins_1 = require("./mixins");
class LeftChatMemberContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'left_chat_member',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
    /** Left chat member */
    get eventMember() {
        return new structures_1.User(this.payload.left_chat_member);
    }
}
exports.LeftChatMemberContext = LeftChatMemberContext;
(0, helpers_1.applyMixins)(LeftChatMemberContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(LeftChatMemberContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            eventMember: context.eventMember
        };
    }
});
