// puregram/lib/contexts/message-auto-delete-timer-changed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageAutoDeleteTimerChangedContext = void 0;
const inspectable_1 = require("inspectable");
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const mixins_1 = require("./mixins");
class MessageAutoDeleteTimerChangedContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'message_auto_delete_timer_changed',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
    /** Message auto delete timer */
    get autoDeleteTimer() {
        return new structures_1.MessageAutoDeleteTimerChanged(this.payload.message_auto_delete_timer_changed);
    }
}
exports.MessageAutoDeleteTimerChangedContext = MessageAutoDeleteTimerChangedContext;
(0, helpers_1.applyMixins)(MessageAutoDeleteTimerChangedContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(MessageAutoDeleteTimerChangedContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            autoDeleteTimer: context.autoDeleteTimer
        };
    }
});
