// puregram/lib/contexts/migrate-to-chat-id.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MigrateToChatIdContext = void 0;
const inspectable_1 = require("inspectable");
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const mixins_1 = require("./mixins");
class MigrateToChatIdContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'migrate_to_chat_id',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
    /** Chat ID */
    get eventId() {
        return this.payload.migrate_to_chat_id;
    }
}
exports.MigrateToChatIdContext = MigrateToChatIdContext;
(0, helpers_1.applyMixins)(MigrateToChatIdContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(MigrateToChatIdContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            eventId: context.eventId
        };
    }
});
