// puregram/lib/contexts/new-chat-title.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewChatTitleContext = void 0;
const inspectable_1 = require("inspectable");
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const mixins_1 = require("./mixins");
class NewChatTitleContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'new_chat_title',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
    /** New chat title */
    get eventTitle() {
        return this.payload.new_chat_title;
    }
}
exports.NewChatTitleContext = NewChatTitleContext;
(0, helpers_1.applyMixins)(NewChatTitleContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(NewChatTitleContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            eventTitle: context.eventTitle
        };
    }
});
