// puregram/lib/contexts/new-chat-members.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewChatMembersContext = void 0;
const inspectable_1 = require("inspectable");
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const mixins_1 = require("./mixins");
class NewChatMembersContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'new_chat_members',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
    /** New chat members */
    get eventMembers() {
        const members = this.payload.new_chat_members;
        return members.map(member => new structures_1.User(member));
    }
}
exports.NewChatMembersContext = NewChatMembersContext;
(0, helpers_1.applyMixins)(NewChatMembersContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(NewChatMembersContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            eventMembers: context.eventMembers
        };
    }
});
