// puregram/lib/contexts/passport-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PassportDataContext = void 0;
const inspectable_1 = require("inspectable");
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const mixins_1 = require("./mixins");
class PassportDataContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'passport_data',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
    /** Telegram Passport data */
    get passportData() {
        return new structures_1.PassportData(this.payload.passport_data);
    }
}
exports.PassportDataContext = PassportDataContext;
(0, helpers_1.applyMixins)(PassportDataContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(PassportDataContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            passportData: context.passportData
        };
    }
});
