// puregram/lib/contexts/pinned-message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PinnedMessageContext = void 0;
const inspectable_1 = require("inspectable");
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const message_1 = require("./message");
const mixins_1 = require("./mixins");
class PinnedMessageContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'pinned_message',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
    /** Pinned message */
    get eventMessage() {
        return new message_1.MessageContext({
            telegram: this.telegram,
            payload: this.payload.pinned_message
        });
    }
}
exports.PinnedMessageContext = PinnedMessageContext;
(0, helpers_1.applyMixins)(PinnedMessageContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(PinnedMessageContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            eventMessage: context.eventMessage
        };
    }
});
