// puregram/lib/contexts/proximity-alert-triggered.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProximityAlertTriggeredContext = void 0;
const inspectable_1 = require("inspectable");
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const mixins_1 = require("./mixins");
class ProximityAlertTriggeredContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'proximity_alert_triggered',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
    /**
     * Service message.
     * A user in the chat triggered another user's proximity alert
     * while sharing Live Location.
     */
    get proximityAlert() {
        return new structures_1.ProximityAlertTriggered(this.payload.proximity_alert_triggered);
    }
}
exports.ProximityAlertTriggeredContext = ProximityAlertTriggeredContext;
(0, helpers_1.applyMixins)(ProximityAlertTriggeredContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(ProximityAlertTriggeredContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            proximityAlert: context.proximityAlert
        };
    }
});
