// puregram/lib/contexts/shipping-query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShippingQueryContext = void 0;
const inspectable_1 = require("inspectable");
const helpers_1 = require("../utils/helpers");
const structures_1 = require("../common/structures");
const context_1 = require("./context");
const mixins_1 = require("./mixins");
class ShippingQueryContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'shipping_query',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
}
exports.ShippingQueryContext = ShippingQueryContext;
(0, helpers_1.applyMixins)(ShippingQueryContext, [structures_1.ShippingQuery, mixins_1.SendMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(ShippingQueryContext, {
    serialize(context) {
        const payload = {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            invoicePayload: context.invoicePayload,
            shippingAddress: context.shippingAddress
        };
        return (0, helpers_1.filterPayload)(payload);
    }
});
