// puregram/lib/contexts/user-shared.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserSharedContext = void 0;
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const inspectable_1 = require("inspectable");
const mixins_1 = require("./mixins");
/** This object contains information about the user whose identifier was shared with the bot using a `KeyboardButtonRequestUser` button. */
class UserSharedContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'user_shared',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
        this.event = this.payload.user_shared;
    }
    /** Identifier of the request */
    get requestId() {
        return this.event.request_id;
    }
    /** Identifier of the shared user. This number may have more than 32 significant bits and some programming languages may have difficulty/silent defects in interpreting it. But it has at most 52 significant bits, so a 64-bit integer or double-precision float type are safe for storing this identifier. The bot may not have access to the user and could be unable to use this identifier, unless the user is already known to the bot by some other means. */
    get sharedUserId() {
        return this.event.user_id;
    }
}
exports.UserSharedContext = UserSharedContext;
(0, helpers_1.applyMixins)(UserSharedContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(UserSharedContext, {
    serialize(context) {
        return {
            requestId: context.requestId,
            sharedUserId: context.sharedUserId
        };
    }
});
