// puregram/lib/contexts/video-chat-scheduled.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VideoChatScheduledContext = void 0;
const inspectable_1 = require("inspectable");
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const mixins_1 = require("./mixins");
class VideoChatScheduledContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'video_chat_scheduled',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
    /** Service message: video chat scheduled */
    get eventScheduled() {
        return new structures_1.VideoChatScheduled(this.payload.video_chat_scheduled);
    }
}
exports.VideoChatScheduledContext = VideoChatScheduledContext;
(0, helpers_1.applyMixins)(VideoChatScheduledContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(VideoChatScheduledContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            eventScheduled: context.eventScheduled
        };
    }
});
