// puregram/lib/contexts/video-chat-participants-invited.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VideoChatParticipantsInvitedContext = void 0;
const inspectable_1 = require("inspectable");
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const mixins_1 = require("./mixins");
class VideoChatParticipantsInvitedContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'video_chat_participants_invited',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
    /** Service message: new participants invited to a video chat */
    get eventParticipantsInvited() {
        return new structures_1.VideoChatParticipantsInvited(this.payload.video_chat_participants_invited);
    }
}
exports.VideoChatParticipantsInvitedContext = VideoChatParticipantsInvitedContext;
(0, helpers_1.applyMixins)(VideoChatParticipantsInvitedContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(VideoChatParticipantsInvitedContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            eventParticipantsInvited: context.eventParticipantsInvited
        };
    }
});
