// puregram/lib/contexts/video-chat-ended.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VideoChatEndedContext = void 0;
const inspectable_1 = require("inspectable");
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const mixins_1 = require("./mixins");
class VideoChatEndedContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'video_chat_ended',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
    /** Service message: video chat ended */
    get eventEnded() {
        return new structures_1.VideoChatEnded(this.payload.video_chat_ended);
    }
}
exports.VideoChatEndedContext = VideoChatEndedContext;
(0, helpers_1.applyMixins)(VideoChatEndedContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(VideoChatEndedContext, {
    serialize(context) {
        return {
            id: context.id,
            from: context.from,
            senderId: context.senderId,
            createdAt: context.createdAt,
            chat: context.chat,
            chatId: context.chatId,
            chatType: context.chatType,
            eventEnded: context.eventEnded
        };
    }
});
