// puregram/lib/contexts/write-access-allowed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WriteAccessAllowedContext = void 0;
const structures_1 = require("../common/structures");
const helpers_1 = require("../utils/helpers");
const context_1 = require("./context");
const inspectable_1 = require("inspectable");
const mixins_1 = require("./mixins");
class WriteAccessAllowedContext extends context_1.Context {
    constructor(options) {
        super({
            telegram: options.telegram,
            updateType: 'write_access_allowed',
            updateId: options.updateId,
            update: options.update
        });
        this.payload = options.payload;
    }
}
exports.WriteAccessAllowedContext = WriteAccessAllowedContext;
(0, helpers_1.applyMixins)(WriteAccessAllowedContext, [structures_1.Message, mixins_1.TargetMixin, mixins_1.SendMixin, mixins_1.NodeMixin, mixins_1.CloneMixin]);
(0, inspectable_1.inspectable)(WriteAccessAllowedContext, {
    serialize(context) {
        return {};
    }
});
