// puregram/lib/common/keyboards/force-reply.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForceReply = void 0;
const inspectable_1 = require("inspectable");
/** Force reply keyboard */
class ForceReply {
    constructor() {
        this.isSelective = false;
    }
    /** Use this parameter if you want to show the keyboard to specific users only */
    selective(selective = true) {
        this.isSelective = selective;
        return this;
    }
    /** The placeholder to be shown in the input field when the keyboard is active */
    setPlaceholder(placeholder) {
        this.placeholder = placeholder;
        return this;
    }
    /** Returns JSON which is compatible with Telegram's `ForceReply` interface */
    toJSON() {
        return {
            force_reply: true,
            input_field_placeholder: this.placeholder,
            selective: this.isSelective
        };
    }
    toString() {
        return JSON.stringify(this);
    }
}
exports.ForceReply = ForceReply;
(0, inspectable_1.inspectable)(ForceReply, {
    serialize(keyboard) {
        return keyboard.toJSON();
    }
});
