// puregram/lib/common/structures/bot-command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BotCommand = void 0;
const inspectable_1 = require("inspectable");
/** This object represents a bot command */
class BotCommand {
    constructor(payload) {
        this.payload = payload;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    /** Text of the command; 1-32 characters. Can contain only lowercase English letters, digits and underscores. */
    get command() {
        return this.payload.command;
    }
    /** Description of the command; 1-256 characters */
    get description() {
        return this.payload.description;
    }
    toJSON() {
        return {
            command: this.command,
            description: this.description
        };
    }
}
exports.BotCommand = BotCommand;
(0, inspectable_1.inspectable)(BotCommand, {
    serialize(struct) {
        return {
            command: struct.command,
            description: struct.description
        };
    }
});
