// puregram/lib/common/structures/callback-query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CallbackQuery = void 0;
const inspectable_1 = require("inspectable");
const helpers_1 = require("../../utils/helpers");
const message_1 = require("./message");
const user_1 = require("./user");
/**
 * This object represents an incoming callback query from a callback button in
 * an inline keyboard. If the button that originated the query was attached to
 * a message sent by the bot, the field message will be present.
 * If the button was attached to a message sent via the bot (in inline mode),
 * the field inline_message_id will be present.
 * Exactly one of the fields `data` or `game_short_name` will be present.
 */
class CallbackQuery {
    constructor(payload) {
        this.payload = payload;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    /** Unique identifier for this query */
    get id() {
        return this.payload.id;
    }
    /** Sender */
    get from() {
        return new user_1.User(this.payload.from);
    }
    /** Sender ID */
    get senderId() {
        return this.from.id;
    }
    /**
     * Message with the callback button that originated the query.
     * Note that message content and message date will not be available
     * if the message is too old
     */
    get message() {
        const { message } = this.payload;
        if (!message) {
            return;
        }
        return new message_1.Message(message);
    }
    /**
     * Identifier of the message sent via the bot in inline mode,
     * that originated the query.
     */
    get inlineMessageId() {
        return this.payload.inline_message_id;
    }
    /**
     * Global identifier, uniquely corresponding to the chat to which the message
     * with the callback button was sent. Useful for high scores in games.
     */
    get chatInstance() {
        return this.payload.chat_instance;
    }
    /**
     * Data associated with the callback button.
     * Be aware that a bad client can send arbitrary data in this field.
     */
    get data() {
        return this.payload.data;
    }
    /**
     * Short name of a Game to be returned,
     * serves as the unique identifier for the game
     */
    get gameShortName() {
        return this.payload.game_short_name;
    }
    toJSON() {
        var _a;
        return {
            id: this.id,
            from: this.from.toJSON(),
            message: (_a = this.message) === null || _a === void 0 ? void 0 : _a.toJSON(),
            inline_message_id: this.inlineMessageId,
            chat_instance: this.chatInstance,
            data: this.data,
            game_short_name: this.gameShortName
        };
    }
}
exports.CallbackQuery = CallbackQuery;
(0, inspectable_1.inspectable)(CallbackQuery, {
    serialize(struct) {
        const payload = {
            id: struct.id,
            from: struct.from,
            message: struct.message,
            inlineMessageId: struct.inlineMessageId,
            chatInstance: struct.chatInstance,
            data: struct.data,
            gameShortName: struct.gameShortName
        };
        return (0, helpers_1.filterPayload)(payload);
    }
});
