// puregram/lib/common/structures/chat-location.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatLocation = void 0;
const inspectable_1 = require("inspectable");
const location_1 = require("./location");
class ChatLocation {
    constructor(payload) {
        this.payload = payload;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    /** The location to which the supergroup is connected. Can't be a live location. */
    get location() {
        return new location_1.Location(this.payload.location);
    }
    /** Location address; `1-64` characters, as defined by the chat owner */
    get address() {
        return this.payload.address;
    }
    toJSON() {
        return {
            address: this.address,
            location: this.location.toJSON()
        };
    }
}
exports.ChatLocation = ChatLocation;
(0, inspectable_1.inspectable)(ChatLocation, {
    serialize(struct) {
        return {
            location: struct.location,
            address: struct.address
        };
    }
});
