// puregram/lib/common/structures/contact.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Contact = void 0;
const inspectable_1 = require("inspectable");
const helpers_1 = require("../../utils/helpers");
/** This object represents a phone contact. */
class Contact {
    constructor(payload) {
        this.payload = payload;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    /** Contact's phone number */
    get phoneNumber() {
        return this.payload.phone_number;
    }
    /** Contact's first name */
    get firstName() {
        return this.payload.first_name;
    }
    /** Contact's last name */
    get lastName() {
        return this.payload.last_name;
    }
    /** Contact's user identifier in Telegram */
    get userId() {
        return this.payload.user_id;
    }
    /** Additional data about the contact in the form of a vCard */
    get vCard() {
        return this.payload.vcard;
    }
    toJSON() {
        return {
            phone_number: this.phoneNumber,
            first_name: this.firstName,
            last_name: this.lastName,
            user_id: this.userId,
            vcard: this.vCard
        };
    }
}
exports.Contact = Contact;
(0, inspectable_1.inspectable)(Contact, {
    serialize(struct) {
        const payload = {
            phoneNumber: struct.phoneNumber,
            firstName: struct.firstName,
            lastName: struct.lastName,
            userId: struct.userId,
            vCard: struct.vCard
        };
        return (0, helpers_1.filterPayload)(payload);
    }
});
