// puregram/lib/common/structures/file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.File = void 0;
const inspectable_1 = require("inspectable");
const helpers_1 = require("../../utils/helpers");
class File {
    constructor(payload) {
        this.payload = payload;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    /**
     * Identifier for this file, which can be used to download or reuse the file
     */
    get fileId() {
        return this.payload.file_id;
    }
    /**
     * Unique identifier for this file, which is supposed to be the same over
     * time and for different bots. Can't be used to download or reuse the file.
     */
    get fileUniqueId() {
        return this.payload.file_unique_id;
    }
    /** File size, if known */
    get fileSize() {
        return this.payload.file_size;
    }
    /**
     * File path.
     * Use `https://api.telegram.org/file/bot<token>/<file_path>` to get the
     * file.
     */
    get filePath() {
        return this.payload.file_path;
    }
    toJSON() {
        return {
            file_id: this.fileId,
            file_unique_id: this.fileUniqueId,
            file_size: this.fileSize,
            file_path: this.filePath
        };
    }
}
exports.File = File;
(0, inspectable_1.inspectable)(File, {
    serialize(struct) {
        const payload = {
            fileId: struct.fileId,
            fileUniqueId: struct.fileUniqueId,
            fileSize: struct.fileSize,
            filePath: struct.filePath
        };
        return (0, helpers_1.filterPayload)(payload);
    }
});
