// puregram/lib/common/structures/forum-topic-created.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForumTopicCreated = void 0;
const inspectable_1 = require("inspectable");
const helpers_1 = require("../../utils/helpers");
/** This object represents a service message about a new forum topic created in the chat. */
class ForumTopicCreated {
    constructor(payload) {
        this.payload = payload;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    /** Name of the topic */
    get name() {
        return this.payload.name;
    }
    /** Color of the topic icon in RGB format */
    get iconColor() {
        return this.payload.icon_color;
    }
    /** Unique identifier of the custom emoji shown as the topic icon */
    get iconCustomEmojiId() {
        return this.payload.icon_custom_emoji_id;
    }
    toJSON() {
        return {
            name: this.name,
            icon_color: this.iconColor,
            icon_custom_emoji_id: this.iconCustomEmojiId
        };
    }
}
exports.ForumTopicCreated = ForumTopicCreated;
(0, inspectable_1.inspectable)(ForumTopicCreated, {
    serialize(struct) {
        const payload = {
            name: struct.name,
            iconColor: struct.iconColor,
            iconCustomEmojiId: struct.iconCustomEmojiId
        };
        return (0, helpers_1.filterPayload)(payload);
    }
});
