// puregram/lib/common/structures/forum-topic-edited.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForumTopicEdited = void 0;
const inspectable_1 = require("inspectable");
const helpers_1 = require("../../utils/helpers");
/** This object represents a service message about an edited forum topic. */
class ForumTopicEdited {
    constructor(payload) {
        this.payload = payload;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    /** New name of the topic, if it was edited */
    get name() {
        return this.payload.name;
    }
    /** New identifier of the custom emoji shown as the topic icon, if it was edited; an empty string if the icon was removed */
    get iconCustomEmojiId() {
        return this.payload.icon_custom_emoji_id;
    }
    toJSON() {
        return {
            name: this.name,
            icon_custom_emoji_id: this.iconCustomEmojiId
        };
    }
}
exports.ForumTopicEdited = ForumTopicEdited;
(0, inspectable_1.inspectable)(ForumTopicEdited, {
    serialize(struct) {
        const payload = {
            name: struct.name,
            iconCustomEmojiId: struct.iconCustomEmojiId
        };
        return (0, helpers_1.filterPayload)(payload);
    }
});
