// puregram/lib/common/structures/menu-button.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MenuButton = void 0;
const inspectable_1 = require("inspectable");
const helpers_1 = require("../../utils/helpers");
const web_app_info_1 = require("./web-app-info");
/** This object describes the bot's menu button in a private chat. */
class MenuButton {
    constructor(payload) {
        this.payload = payload;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    /** Type of the button */
    get type() {
        return this.payload.type;
    }
    /** Text on the button */
    get text() {
        return this.payload.text;
    }
    /**
     * Description of the Web App that will be launched when the user presses the button.
     * The Web App will be able to send an arbitrary message on behalf of the user
     * using the method `answerWebAppQuery`.
     */
    get webApp() {
        const { web_app } = this.payload;
        if (!web_app) {
            return;
        }
        return new web_app_info_1.WebAppInfo(web_app);
    }
    toJSON() {
        var _a;
        // @ts-expect-error typescript yells at type for no reason :smh:
        return {
            type: this.type,
            text: this.text,
            web_app: (_a = this.webApp) === null || _a === void 0 ? void 0 : _a.toJSON()
        };
    }
}
exports.MenuButton = MenuButton;
(0, inspectable_1.inspectable)(MenuButton, {
    serialize(struct) {
        const payload = {
            type: struct.type,
            text: struct.text,
            webApp: struct.webApp
        };
        return (0, helpers_1.filterPayload)(payload);
    }
});
