// puregram/lib/common/structures/order-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderInfo = void 0;
const inspectable_1 = require("inspectable");
const helpers_1 = require("../../utils/helpers");
const shipping_address_1 = require("./shipping-address");
/** This object represents information about an order. */
class OrderInfo {
    constructor(payload) {
        this.payload = payload;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    /** User name */
    get name() {
        return this.payload.name;
    }
    /** User's phone number */
    get phoneNumber() {
        return this.payload.phone_number;
    }
    /** User email */
    get email() {
        return this.payload.email;
    }
    /** User shipping address */
    get shippingAddress() {
        const { shipping_address } = this.payload;
        if (!shipping_address) {
            return;
        }
        return new shipping_address_1.ShippingAddress(shipping_address);
    }
    toJSON() {
        var _a;
        return {
            name: this.name,
            phone_number: this.phoneNumber,
            email: this.email,
            shipping_address: (_a = this.shippingAddress) === null || _a === void 0 ? void 0 : _a.toJSON()
        };
    }
}
exports.OrderInfo = OrderInfo;
(0, inspectable_1.inspectable)(OrderInfo, {
    serialize(struct) {
        const payload = {
            name: struct.name,
            phoneNumber: struct.phoneNumber,
            email: struct.email,
            shippingAddress: struct.shippingAddress
        };
        return (0, helpers_1.filterPayload)(payload);
    }
});
