// puregram/lib/common/structures/poll-answer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PollAnswer = void 0;
const inspectable_1 = require("inspectable");
const user_1 = require("./user");
/** This object represents an answer of a user in a non-anonymous poll. */
class PollAnswer {
    constructor(payload) {
        this.payload = payload;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    /** Unique poll identifier */
    get pollId() {
        return this.payload.poll_id;
    }
    /** The user, who changed the answer to the poll */
    get user() {
        return new user_1.User(this.payload.user);
    }
    /** Sender ID */
    get senderId() {
        return this.user.id;
    }
    /**
     * 0-based identifiers of answer options, chosen by the user.
     * May be empty if the user retracted their vote.
     */
    get optionIds() {
        return this.payload.option_ids;
    }
    toJSON() {
        return {
            poll_id: this.pollId,
            user: this.user.toJSON(),
            sender_id: this.senderId,
            option_ids: this.optionIds
        };
    }
}
exports.PollAnswer = PollAnswer;
(0, inspectable_1.inspectable)(PollAnswer, {
    serialize(struct) {
        return {
            pollId: struct.pollId,
            user: struct.user,
            senderId: struct.senderId,
            optionIds: struct.optionIds
        };
    }
});
