// puregram/lib/common/structures/proximity-alert-triggered.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProximityAlertTriggered = void 0;
const inspectable_1 = require("inspectable");
const user_1 = require("./user");
/**
 * This object represents the content of a service message,
 * sent whenever a user in the chat triggers a proximity alert set by another user.
 */
class ProximityAlertTriggered {
    constructor(payload) {
        this.payload = payload;
    }
    /** User that triggered the alert */
    get traveler() {
        return new user_1.User(this.payload.traveler);
    }
    /** User that set the alert */
    get watcher() {
        return new user_1.User(this.payload.watcher);
    }
    /** The distance between the users */
    get distance() {
        return this.payload.distance;
    }
    toJSON() {
        return {
            traveler: this.traveler.toJSON(),
            watcher: this.watcher.toJSON(),
            distance: this.distance
        };
    }
}
exports.ProximityAlertTriggered = ProximityAlertTriggered;
(0, inspectable_1.inspectable)(ProximityAlertTriggered, {
    serialize(struct) {
        return {
            traveler: struct.traveler,
            watcher: struct.watcher,
            distance: struct.distance
        };
    }
});
