// puregram/lib/common/structures/sent-web-app-message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SentWebAppMessage = void 0;
const inspectable_1 = require("inspectable");
const helpers_1 = require("../../utils/helpers");
/** Contains information about an inline message sent by a Web App on behalf of a user. */
class SentWebAppMessage {
    constructor(payload) {
        this.payload = payload;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    /**
     * Identifier of the sent inline message.
     *
     * Available only if there is an inline keyboard attached to the message.
     */
    get inlineMessageId() {
        return this.payload.inline_message_id;
    }
    toJSON() {
        return {
            inline_message_id: this.inlineMessageId
        };
    }
}
exports.SentWebAppMessage = SentWebAppMessage;
(0, inspectable_1.inspectable)(SentWebAppMessage, {
    serialize(struct) {
        const payload = {
            inlineMessageId: struct.inlineMessageId
        };
        return (0, helpers_1.filterPayload)(payload);
    }
});
