// puregram/lib/common/structures/shipping-query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShippingQuery = void 0;
const inspectable_1 = require("inspectable");
const user_1 = require("./user");
const shipping_address_1 = require("./shipping-address");
/** This object contains information about an incoming shipping query. */
class ShippingQuery {
    constructor(payload) {
        this.payload = payload;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    /** Unique query identifier */
    get id() {
        return this.payload.id;
    }
    /** User who sent the query */
    get from() {
        return new user_1.User(this.payload.from);
    }
    /** Sender ID */
    get senderId() {
        return this.from.id;
    }
    /** Bot specified invoice payload */
    get invoicePayload() {
        return this.payload.invoice_payload;
    }
    /** User specified shipping address */
    get shippingAddress() {
        return new shipping_address_1.ShippingAddress(this.payload.shipping_address);
    }
    toJSON() {
        return {
            id: this.id,
            from: this.from.toJSON(),
            invoice_payload: this.invoicePayload,
            shipping_address: this.shippingAddress.toJSON()
        };
    }
}
exports.ShippingQuery = ShippingQuery;
(0, inspectable_1.inspectable)(ShippingQuery, {
    serialize(struct) {
        return {
            id: struct.id,
            from: struct.from,
            senderId: struct.senderId,
            invoicePayload: struct.invoicePayload,
            shippingAddress: struct.shippingAddress
        };
    }
});
