// puregram/lib/common/structures/successful-payment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SuccessfulPayment = void 0;
const inspectable_1 = require("inspectable");
const helpers_1 = require("../../utils/helpers");
const order_info_1 = require("./order-info");
/** This object contains basic information about a successful payment. */
class SuccessfulPayment {
    constructor(payload) {
        this.payload = payload;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    /** Three-letter ISO 4217 currency code */
    get currency() {
        return this.payload.currency;
    }
    /**
     * Total price in the smallest units of the currency
     * (integer, not float/double). For example, for a price of
     * `US$ 1.45` pass `amount = 145`. See the `exp` parameter in
     * [currencies.json](https://core.telegram.org/bots/payments/currencies.json),
     * it shows the number of digits past the decimal point for each currency
     * (2 for the majority of currencies).
     */
    get totalAmount() {
        return this.payload.total_amount;
    }
    /** Bot specified invoice payload */
    get invoicePayload() {
        return this.payload.invoice_payload;
    }
    /** Identifier of the shipping option chosen by the user */
    get shippingOptionId() {
        return this.payload.shipping_option_id;
    }
    /** Order info provided by the user */
    get orderInfo() {
        const { order_info } = this.payload;
        if (!order_info) {
            return;
        }
        return new order_info_1.OrderInfo(order_info);
    }
    /** Telegram payment identifier */
    get telegramPaymentChargeId() {
        return this.payload.telegram_payment_charge_id;
    }
    /** Provider payment identifier */
    get providerPaymentChargeId() {
        return this.payload.provider_payment_charge_id;
    }
    toJSON() {
        return {
            currency: this.currency,
            total_amount: this.totalAmount,
            invoice_payload: this.invoicePayload,
            shipping_option_id: this.shippingOptionId,
            order_info: this.orderInfo,
            telegram_payment_charge_id: this.telegramPaymentChargeId,
            provider_payment_charge_id: this.providerPaymentChargeId
        };
    }
}
exports.SuccessfulPayment = SuccessfulPayment;
(0, inspectable_1.inspectable)(SuccessfulPayment, {
    serialize(struct) {
        const payload = {
            currency: struct.currency,
            totalAmount: struct.totalAmount,
            invoicePayload: struct.invoicePayload,
            shippingOptionId: struct.shippingOptionId,
            orderInfo: struct.orderInfo,
            telegramPaymentChargeId: struct.telegramPaymentChargeId,
            providerPaymentChargeId: struct.providerPaymentChargeId
        };
        return (0, helpers_1.filterPayload)(payload);
    }
});
