// puregram/lib/common/structures/user-profile-photos.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserProfilePhotos = void 0;
const inspectable_1 = require("inspectable");
const photo_size_1 = require("./photo-size");
/** This object represent a user's profile pictures. */
class UserProfilePhotos {
    constructor(payload) {
        this.payload = payload;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    /** Total number of profile pictures the target user has */
    get totalCount() {
        return this.payload.total_count;
    }
    /** Requested profile pictures (in up to 4 sizes each) */
    get photos() {
        const { photos } = this.payload;
        if (!photos.length) {
            return;
        }
        return photos.map(row => row.map(element => new photo_size_1.PhotoSize(element)));
    }
    toJSON() {
        var _a, _b;
        return {
            photos: (_b = (_a = this.photos) === null || _a === void 0 ? void 0 : _a.map(row => row.map(size => size.toJSON()))) !== null && _b !== void 0 ? _b : [],
            total_count: this.totalCount
        };
    }
}
exports.UserProfilePhotos = UserProfilePhotos;
(0, inspectable_1.inspectable)(UserProfilePhotos, {
    serialize(struct) {
        return {
            totalCount: struct.totalCount,
            photos: struct.photos
        };
    }
});
