// puregram/lib/common/structures/web-app-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebAppData = void 0;
const inspectable_1 = require("inspectable");
/** Contains data sent from a Web App to the bot. */
class WebAppData {
    constructor(payload) {
        this.payload = payload;
    }
    /** The data. Be aware that a bad client can send arbitrary data in this field. */
    get data() {
        return this.payload.data;
    }
    /**
     * Text of the `web_app` keyboard button, from which the Web App was opened.
     * Be aware that a bad client can send arbitrary data in this field.
     */
    get buttonText() {
        return this.payload.button_text;
    }
    toJSON() {
        return {
            data: this.data,
            button_text: this.buttonText
        };
    }
}
exports.WebAppData = WebAppData;
(0, inspectable_1.inspectable)(WebAppData, {
    serialize(struct) {
        return {
            data: struct.data,
            buttonText: struct.buttonText
        };
    }
});
