// puregram/lib/contexts/mixins/target.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TargetMixin = void 0;
const structures_1 = require("../../common/structures");
const enums_1 = require("../../types/enums");
/** This object represents a mixin which has sender data (e.g. `senderId`, `from` etc.) */
class TargetMixin {
    /** Checks if the instance has `from` and `senderId` properties */
    hasFrom() {
        return this.payload.from !== undefined;
    }
    /** Sender, empty for messages sent to channels */
    get from() {
        const { from } = this.payload;
        if (!from) {
            return;
        }
        return new structures_1.User(from);
    }
    /** Checks if the instance has `senderChat` property */
    hasSenderChat() {
        return this.payload.sender_chat !== undefined;
    }
    /**
     * Sender of the message, sent on behalf of a chat.
     * The channel itself for channel messages.
     * The supergroup itself for messages from anonymous group administrators.
     * The linked channel for messages automatically forwarded to the discussion group
     */
    get senderChat() {
        const { sender_chat } = this.payload;
        if (!sender_chat) {
            return;
        }
        return new structures_1.Chat(sender_chat);
    }
    /** Conversation the message belongs to */
    get chat() {
        return new structures_1.Chat(this.payload.chat);
    }
    /** Sender's ID */
    get senderId() {
        var _a;
        return (_a = this.from) === null || _a === void 0 ? void 0 : _a.id;
    }
    /** Chat ID */
    get chatId() {
        return this.chat.id;
    }
    /** Chat type */
    get chatType() {
        return this.chat.type;
    }
    /** Is this chat a private one? */
    isPM() {
        return this.chatType === enums_1.ChatType.Private;
    }
    /** Is this chat a group? */
    isGroup() {
        return this.chatType === enums_1.ChatType.Group;
    }
    /** Is this chat a supergroup? */
    isSupergroup() {
        return this.chatType === enums_1.ChatType.Supergroup;
    }
    /** Is this chat a channel? */
    isChannel() {
        return this.chatType === enums_1.ChatType.Channel;
    }
}
exports.TargetMixin = TargetMixin;
