// puregram/lib/types/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PollType = exports.ChatType = exports.UpdateType = exports.EntityType = exports.AttachmentType = void 0;
var AttachmentType;
(function (AttachmentType) {
    AttachmentType["Animation"] = "animation";
    AttachmentType["Audio"] = "audio";
    AttachmentType["Contact"] = "contact";
    AttachmentType["Document"] = "document";
    AttachmentType["Location"] = "location";
    AttachmentType["Photo"] = "photo";
    AttachmentType["Poll"] = "poll";
    AttachmentType["Sticker"] = "sticker";
    AttachmentType["Venue"] = "venue";
    AttachmentType["VideoNote"] = "video_note";
    AttachmentType["Video"] = "video";
    AttachmentType["Voice"] = "voice";
})(AttachmentType = exports.AttachmentType || (exports.AttachmentType = {}));
var EntityType;
(function (EntityType) {
    EntityType["Mention"] = "mention";
    EntityType["Hashtag"] = "hashtag";
    EntityType["Cashtag"] = "cashtag";
    EntityType["BotCommand"] = "bot_command";
    EntityType["Url"] = "url";
    EntityType["Email"] = "email";
    EntityType["PhoneNumber"] = "phone_number";
    EntityType["Bold"] = "bold";
    EntityType["Italic"] = "italic";
    EntityType["Underline"] = "underline";
    EntityType["Strikethrough"] = "strikethrough";
    EntityType["Code"] = "code";
    EntityType["Pre"] = "pre";
    EntityType["TextLink"] = "text_link";
    EntityType["TextMention"] = "text_mention";
})(EntityType = exports.EntityType || (exports.EntityType = {}));
var UpdateType;
(function (UpdateType) {
    UpdateType["CallbackQuery"] = "callback_query";
    UpdateType["ChannelChatCreated"] = "channel_chat_created";
    UpdateType["ChatJoinRequest"] = "chat_join_request";
    UpdateType["ChatMember"] = "chat_member";
    UpdateType["MyChatMember"] = "my_chat_member";
    UpdateType["ChosenInlineResult"] = "chosen_inline_result";
    UpdateType["DeleteChatPhoto"] = "delete_chat_photo";
    UpdateType["GroupChatCreated"] = "group_chat_created";
    UpdateType["InlineQuery"] = "inline_query";
    UpdateType["Invoice"] = "invoice";
    UpdateType["LeftChatMember"] = "left_chat_member";
    UpdateType["Location"] = "location";
    UpdateType["MessageAutoDeleteTimerChanged"] = "message_auto_delete_timer_changed";
    UpdateType["Message"] = "message";
    UpdateType["ChannelPost"] = "channel_post";
    UpdateType["EditedMessage"] = "edited_message";
    UpdateType["EditedChannelPost"] = "edited_channel_post";
    UpdateType["MigrateFromChatId"] = "migrate_from_chat_id";
    UpdateType["MigrateToChatId"] = "migrate_to_chat_id";
    UpdateType["NewChatMembers"] = "new_chat_members";
    UpdateType["NewChatPhoto"] = "new_chat_photo";
    UpdateType["NewChatTitle"] = "new_chat_title";
    UpdateType["PassportData"] = "passport_data";
    UpdateType["PinnedMessage"] = "pinned_message";
    UpdateType["PollAnswer"] = "poll_answer";
    UpdateType["Poll"] = "poll";
    UpdateType["PreCheckoutQuery"] = "pre_checkout_query";
    UpdateType["ProximityAlertTriggered"] = "proximity_alert_triggered";
    UpdateType["WriteAccessAllowed"] = "write_access_allowed";
    UpdateType["ForumTopicCreated"] = "forum_topic_created";
    UpdateType["ForumTopicEdited"] = "forum_topic_edited";
    UpdateType["ForumTopicClosed"] = "forum_topic_closed";
    UpdateType["ForumTopicReopened"] = "forum_topic_reopened";
    UpdateType["GeneralForumTopicHidden"] = "general_forum_topic_hidden";
    UpdateType["GeneralForumTopicUnhidden"] = "general_forum_topic_unhidden";
    UpdateType["ShippingQuery"] = "shipping_query";
    UpdateType["SuccessfulPayment"] = "successful_payment";
    UpdateType["UserShared"] = "user_shared";
    UpdateType["ChatShared"] = "chat_shared";
    UpdateType["SupergroupChatCreated"] = "supergroup_chat_created";
    UpdateType["VideoChatEnded"] = "video_chat_ended";
    UpdateType["VideoChatParticipantsInvited"] = "video_chat_participants_invited";
    UpdateType["VideoChatScheduled"] = "video_chat_scheduled";
    UpdateType["VideoChatStarted"] = "video_chat_started";
    UpdateType["WebAppData"] = "web_app_data";
    UpdateType["ServiceMessage"] = "service_message";
})(UpdateType = exports.UpdateType || (exports.UpdateType = {}));
var ChatType;
(function (ChatType) {
    ChatType["Private"] = "private";
    ChatType["Group"] = "group";
    ChatType["Supergroup"] = "supergroup";
    ChatType["Channel"] = "channel";
})(ChatType = exports.ChatType || (exports.ChatType = {}));
var PollType;
(function (PollType) {
    PollType["Regular"] = "regular";
    PollType["Quiz"] = "quiz";
})(PollType = exports.PollType || (exports.PollType = {}));
